### PR TITLE
PC Relate with Persist annotations [PCRI-3]

### PR DESCRIPTION
**Pull request by danking**
The third in a series of PCRelate Improvements. Depends on #2249 

Sprinkling `cache` on any RDD which is used more than once dramatically improved runtime. On a benchmark program (included below) PCRelate took 40 seconds with four cores on 1000 samples and 10,000 variants.

```
from hail import *
from timeit import default_timer as timer

hc = HailContext()
vds = hc.balding_nichols_model(20, 1000, 10000).repartition(10).persist()
vds.count()

start = timer()
vds.pc_relate(5, 0.01, min_kinship=0.1, desire="all").count()
end = timer()
print(end-start)
```

### CODE REVIEW

File: src/main/scala/is/hail/methods/PCRelate.scala [hunk 1]
```diff
@@ -243,16 +247,16 @@ class PCRelate(maf: Double, blockSize: Int) extends Serializable {
...
-    val homalt = dm.map2 { (g, mu) =>
+    val homalt = (dm.map2 { (g, mu) =>
       if (badgt(g) || badmu(mu) || g != 2.0) 0.0 else 1.0
-    } (g, mu)
-    val homref = dm.map2 { (g, mu) =>
+    } (g, mu)).cache()
+    val homref = (dm.map2 { (g, mu) =>
       if (badgt(g) || badmu(mu) || g != 0.0) 0.0 else 1.0
-    } (g, mu)
+    } (g, mu)).cache
```
Comment (jigold) [inline src/main/scala/is/hail/methods/PCRelate.scala]: choose one of `cache` versus `cache()`.

File: src/main/scala/is/hail/methods/PCRelate.scala [hunk 2]
```diff
@@ -190,12 +190,16 @@ class PCRelate(maf: Double, blockSize: Int) extends Serializable {
     mu <= maf || mu >= antimaf || mu <= 0.0 || mu >= 1.0
   def badgt(gt: Double): Boolean =
     gt != 0.0 && gt != 1.0 && gt != 2.0
+  private def gram(m: M): M = {
```
Comment (jigold) [inline src/main/scala/is/hail/methods/PCRelate.scala]: I prefer spaces between method definitions. Your call.

File: src/main/scala/is/hail/methods/PCRelate.scala [hunk 3]
```diff
@@ -168,28 +190,39 @@ class PCRelate(maf: Double, blockSize: Int) extends Serializable {
...
+    val mu = this.mu(g, pcs).cache()
+    val blockedG = dm.from(g, blockSize, blockSize).cache()
 
-    val variance = dm.map2 {
-      case (g, mu) =>
-        if (badgt(g) || badmu(mu))
-          0.0
-        else
-          mu * (1.0 - mu)
+    val variance = dm.map2 { (g, mu) =>
```
Comment (jigold) [inline src/main/scala/is/hail/methods/PCRelate.scala]: Should the `variance` be cached?

**Comment by danking** [comment]
I've addressed the comments, but I can't run the benchmark anymore, I get weird errors about partitions being empty that I expected to be non-empty. I will continue to investigate.

**Comment by danking** [comment]
Benchmark is now like 80s. Not sure why it got slower from before.

**Comment by danking** [comment]
I'll re-open when #2270 lands because this builds on that.

**Pull request closed without merging**